packages/flutter_tools/lib/src/windows/windows_version_validator.dart

```diff
@@ -111,3 +126,121 @@ class ProcessLister {
+    IntPtr hKey,
+    Pointer<Utf16> lpSubKey,
+    Int32 ulOptions,
+    Int32 samDesired,
+    Pointer<IntPtr> phkResult);
+typedef RegOpenKeyExDart = int Function(
+    int hKey,
+    Pointer<Utf16> lpSubKey,
+    int ulOptions,
+    int samDesired,
+    Pointer<IntPtr> phkResult);
+typedef RegQueryValueExNative = Int32 Function(
+    IntPtr hKey,
+    Pointer<Utf16> lpValueName,
+    Pointer<Int32> lpReserved,
+    Pointer<Int32> lpType,
+    Pointer<Void> lpData,
+    Pointer<Int32> lpcbData);
+typedef RegQueryValueExDart = int Function(
+    int hKey,
+    Pointer<Utf16> lpValueName,
+    Pointer<Int32> lpReserved,
+    Pointer<Int32> lpType,
+    Pointer<Void> lpData,
+    Pointer<Int32> lpcbData);
+
+typedef RegCloseKeyNative = Int32 Function(IntPtr hKey);
+typedef RegCloseKeyDart = int Function(int hKey);
+
+base class KeyValueBasicInformation extends Struct {
+  @Uint32()
+  external int titleIndex;
+  @Uint32()
+  external int type;
+  @Uint32()
+  external int nameLength;
+  @Uint32()
+  external int dataLength;
+}
+
+const int HKEY_LOCAL_MACHINE = 0x80000002;
+const int KEY_READ = 0x20019;
+const int REG_SZ = 1;
+const int REG_DWORD = 4;
+const String CURRENT_VERSION_KEY = r'SOFTWARE\Microsoft\Windows NT\CurrentVersion';
+
+String? readRegistryValue(int hKey, String subKey, String valueName) {
+  final DynamicLibrary advapi32 = DynamicLibrary.open('advapi32.dll');
+  final RegOpenKeyExDart regOpenKeyEx = advapi32.lookupFunction<RegOpenKeyExNative, RegOpenKeyExDart>('RegOpenKeyExW');
+  final RegQueryValueExDart regQueryValueEx = advapi32.lookupFunction<RegQueryValueExNative, RegQueryValueExDart>('RegQueryValueExW');
+  final RegCloseKeyDart regCloseKey = advapi32.lookupFunction<RegCloseKeyNative, RegCloseKeyDart>('RegCloseKey');
+
+  final Pointer<Utf16> subKeyPtr = subKey.toNativeUtf16();
+  final Pointer<Utf16> valueNamePtr = valueName.toNativeUtf16();
+  final Pointer<IntPtr> hSubKeyPtr = calloc<IntPtr>();
+
+  final int openResult = regOpenKeyEx(HKEY_LOCAL_MACHINE, subKeyPtr, 0, KEY_READ, hSubKeyPtr);
+  if (openResult != 0) {
+    calloc.free(hSubKeyPtr);
+    return null;
+  }
+
+  final int hSubKey = hSubKeyPtr.value;
+
+  final Pointer<Int32> dataSizePtr = calloc<Int32>();
+  final Pointer<Int32> typePtr = calloc<Int32>();
+  regQueryValueEx(hSubKey, valueNamePtr, nullptr, typePtr, nullptr, dataSizePtr);
+  final int dataSize = dataSizePtr.value;
+
+  final Pointer<Uint8> dataPtr = calloc<Uint8>(dataSize);
+  regQueryValueEx(hSubKey, valueNamePtr, nullptr, typePtr, dataPtr.cast(), dataSizePtr);
+
+  String? result;
+  if (typePtr.value == REG_SZ) {
+    result = dataPtr.cast<Utf16>().toDartString(length: dataSize ~/ 2);
+  } else if (typePtr.value == REG_DWORD) {
+    result = dataPtr.cast<Uint32>().value.toString();
+  }
+
+  regCloseKey(hSubKey);
+  calloc.free(subKeyPtr);
+  calloc.free(valueNamePtr);
+  calloc.free(hSubKeyPtr);
+  calloc.free(dataSizePtr);
+  calloc.free(typePtr);
+  calloc.free(dataPtr);
```

Comment:
(entirely optional suggestion, feel free to ignore)

```suggestion
  calloc
    ..free(subKeyPtr)
    ..free(valueNamePtr)
    ..free(hSubKeyPtr)
    ..free(dataSizePtr)
    ..free(typePtr)
    ..free(dataPtr);
```